storage/replica.go

```diff
@@ -2594,6 +2535,81 @@ func (r *Replica) processRaftCommand(
 	return pErr
 }
 
+func (r *Replica) acquireSplitMergeLock(ba roachpb.BatchRequest) func(pErr *roachpb.Error) {
+	arg, ok := ba.GetArg(roachpb.EndTransaction)
+	if !ok {
+		return nil
+	}
+	ict := arg.(*roachpb.EndTransactionRequest).InternalCommitTrigger
+	if unlock := r.acquireSplitLock(ict.GetSplitTrigger()); unlock != nil {
+		return unlock
+	}
+	return r.acquireMergeLock(ict.GetMergeTrigger())
+}
+
+func (r *Replica) acquireSplitLock(split *roachpb.SplitTrigger) func(pErr *roachpb.Error) {
+	if split == nil {
+		return nil
+	}
+
+	rightRng, err := r.store.getOrCreateReplica(split.RightDesc.RangeID, 0, nil)
+	if err != nil {
+		return nil
+	}
+
+	if rightRng.IsInitialized() {
+		// TODO(peter,tschottdorf): TestStoreRangeSplitRaceUninitializedRHS sets
+		// up a situation where one of the stores has already processed the
+		// split. We can see this by noticing that there is an initialized
+		// replica for the right hand side. This seems funky, but the magic in
+		// TestStoreRangeSplitRaceUninitializedRHS is too advanced for me to
+		// deciper. Avoiding locking (and re-initializing) the right hand side
+		// range allows the batch to error when Replica.checkCmdHeader determines
+		// that one of the requests in the batch is trying to operate on keys
+		// outside of the Replica's key range. This results in an "error
+		// executing raft command:" message that can be seen by enabling
+		// -vmodule=replica=1.
+		rightRng.raftMu.Unlock()
+		return nil
+	}
+
+	return func(pErr *roachpb.Error) {
+		if pErr != nil {
+			// An error occurred during processing of the split. Mark the RHS
+			// destroyed and remove it from the replica's map as it is likely in an
+			// inconsistent state. One reason this can occur is when concurrent
+			// splits on the same key are executed.
```

Comment:
I think it's unlikely to be inconsistent - concurrent splits will fail in EndTransaction before starting the splitTrigger, so nothing touches the RHS replica. And marking the replica destroyed is a big hammer; we won't be able to recover until rebalancing has moved things around. 

I think if we mark one side of the split as corrupt/destroyed, we should do it to both. So maybe this should check for a ReplicaCorruptionError instead of destroying the RHS on all replicas (and splitTrigger should be modified to return ReplicaCorruptionError when necessary).